core/src/main/java/io/questdb/cairo/sql/PageFrameMemoryPool.java

```diff
@@ -39,13 +50,63 @@
  * This pool is thread-unsafe as it may hold navigated Parquet partition data,
  * so it shouldn't be shared between multiple threads.
  */
-// TODO: add LRU cache for multiple frames
-public class PageFrameMemoryPool implements QuietCloseable {
-    private final PageFrameMemoryImpl frameMemory = new PageFrameMemoryImpl();
+public class PageFrameMemoryPool implements QuietCloseable, Mutable {
+    private static final byte FRAME_MEMORY_MASK = 1 << 2;
+    private static final byte RECORD_A_MASK = 1;
+    private static final byte RECORD_B_MASK = 1 << 1;
+    // LRU cache (most recently used buffers are to the right)
+    private final ObjList<ParquetBuffers> cachedParquetBuffers;
+    private final PageFrameMemoryImpl frameMemory;
+    private final ObjList<ParquetBuffers> freeParquetBuffers;
+    // Contains parquet to query column index mapping.
+    private final IntList fromParquetColumnIndexes;
+    private final int parquetCacheSize;
+    // Contains [parquet_column_index, column_type] pairs.
+    private final DirectIntList parquetColumns;
+    private final PartitionDecoder parquetDecoder;
+    // Contains table reader to parquet column index mapping.
+    private final IntList toParquetColumnIndexes;
     private PageFrameAddressCache addressCache;
 
+    public PageFrameMemoryPool(int parquetCacheSize) {
+        try {
+            this.parquetCacheSize = parquetCacheSize;
+            cachedParquetBuffers = new ObjList<>(parquetCacheSize);
+            freeParquetBuffers = new ObjList<>(parquetCacheSize);
+            for (int i = 0; i < parquetCacheSize; i++) {
+                freeParquetBuffers.add(new ParquetBuffers());
+            }
+            frameMemory = new PageFrameMemoryImpl();
+            toParquetColumnIndexes = new IntList(16);
+            fromParquetColumnIndexes = new IntList(16);
+            parquetColumns = new DirectIntList(32, MemoryTag.NATIVE_DEFAULT);
+            parquetDecoder = new PartitionDecoder();
```

Comment:
nit: this is a confusing pattern, decoder is freed on clear() and then again on close(). Right now there is double-close protection, we should work around decoder.clear() instead. A todo: for now should suffice

Reply:
`close()` is called when, for instance, we catch an exception and close the record cursor. So, it should release all held resources. `clear()` is used when we reset a `PageFrameReduceTask` - in this case we want to reset the decoder, yet retain `parquetColumns` list. Potentially we could close the pool in `PageFrameReduceTask`, but then we might be allocating `parquetColumns` redundantly.